rfcs/webtransport_h3_test_server.md

```diff
@@ -0,0 +1,88 @@
+# RFC 85: WebTransport over HTTP/3 Test Server
+
+## Summary
+
+Add a [WebTransport over HTTP/3](https://datatracker.ietf.org/doc/html/draft-ietf-webtrans-http3-01) server to wpt. The server is used to test the [WebTransport Web APIs](https://w3c.github.io/webtransport/).
+
+The server will use a new [file name flag](https://web-platform-tests.org/writing-tests/file-names.html) `.webtransport` (`.webtransport.py`) for handlers. Handlers are python scripts which are used to respond WebTransport events.
```

Comment:
We may want to have handlers for WebTransport over HTTP/2 and map the same handler directory. How about adding .h3 to the script? The suffix will be : `.webtransport.h3.py` `.wt.h3.py`.

For the context, we are planning to have some tests as "generic" tests, running the same tests for WebTrasnport over HTTP/3 and WebTransport over HTTP/2. For example, the reliable streams behaviors are very similar with each other. We can write something like:

```
// This can be WebTransport over HTTP/3 or WebTransport
// over HTTP/2, depending on `server`.
const wt = new WebTransport(`${server}/echo`);
const stream = await wt.createBidirectionalStream();
// Test `stream`.
...
```

and we want to have `echo.wt.h3.py` and `echo.wt.h2.py`.

Reply:
Thank you for the context. `wt.h3.py` and `wt.h2.py` look reasonable to me. I guess the content of handlers would be the same for most cases so we might also want `wt.any.py`.

Reply:
@jgraham FYI

Reply:
This still doesn't make sense to me :) There's no mechanism in wpt to ensure that something with `.h2` in the name is loaded over HTTP2 or that `.h3` in the name is loaded over HTTP3. Those are just different servers running on different ports. The only thing that cares about the flags is the test manifest generation and, indirectly, the test runner which uses the manifest data to decide which server to use to load the initial HTML test file in the browser.

I can see that the `.webtransport` flag could be used to decide which Python API should be used (although until now we just check which functions exist in the file rather than using a specific naming convention). But `echo.wt.py` is something that can be loaded via any supported HTTP version at the discretion of the HTML test, just by specifying the required server address.

I suppose `.any.js` tests look like they're an exception to this rule about only putting flags in the HTML test filename, but that's because we're (effectively) autogenerating the initial HTML file from the js file. If the plan is to autogenerate tests from just the handler then it would make sense to put the filename flags into the handler file, but I don't think that's correct?

Reply:
There are multiple points to discuss I think.

 1. Where is the document root for the WebTransport over HTTP/3 server?
 2. Should we have the name completion functionality (e.g., "echo" => "echo.wt.h3.py")?
 3. Should we have a convention for WebSocket over HTTP/3 script (e.g., .webstransport, .wt, .wt.h3, .wth3)?

You answers for the questions are:

 1. Where is the document root for the WebTransport over HTTP/3 server?
    The document root of wptserve.
 2. Should we have the name completion functionality (e.g., "echo" => "echo.wt.h3.py")?
    No, for simplicity.
 3. Should we have a convention for WebSocket over HTTP/3 script (e.g., `.webstransport`, `.wt`, `.wt.h3`, `.wth3`)?
   Maybe, fine with `.wt` but not fine with `.wt.h3`.

Is my understanding correct?

Reply:
Yes, I think that's correct.  terms of filename flags, I think we should only use them where they're functional, so that if you get them wrong the test doesn't work. So the question for the `.webtransport` flag on `.py` files is "is that required for the file to be handled correctly". As I understand the proposal, that would be required for the server to identify how to handle the file, so it does make sense. If we need to know whether something is a webtransport handler before the client tries to use it then a filename flag required, if we don't we could introspect the file when it's first loaded like we do for HTTP/2 handlers. But that's a conversation about tradeoffs rather than a hard requirement.

Reply:
I'm fine with the first two answers, but for the last I still want to have at least a convention to distinguish them from normal HTTP handlers, and from each other (WebTransport over HTTP/2 and WebTransport over HTTP/3).

Reply:
I like the way all the WPT WebSocket handlers are organised in a single directory. For comparison, in the Blink WebSocket tests they are scattered among the html files and it's much harder to keep track of them.

We can still put the handlers in a separate directory even if the document root is the same as wptserve.

**Benefits of using the same document root:**
* The handlers are easy to find if the URL contains the directory name.
* Occasionally we've found it useful to do a normal HTTP fetch to the WebSocket server in a test. If it turns out we need to do this in a WebTransport test it would be easier if all the files were already visible under the document root.

**Disadvantages of using the same document root:**
* Every handler URL ends up having the same boilerplate path in it.
* The server will try to load random Python files if the URL points to them, which could have weird side effects.
* The principle of least privilege implies we should make as little as possible be visible from the WebTransport server to reduce the attack surface.

Reply:
The current proposal doesn't require flag names to be functional. I feel we can submit a separate RFC for HTTP/2 vs HTTP/3. I think `.py` is sufficient for now and it's not difficult to add/modify handler names as needed.

Regarding discoverability, I think using the same root as wpt makes sense, and I agree with Adam that it's nice to put handlers in a single directory. It might be cumbersome having the boilerplate paths but I guess that's acceptable compromise.

Updated the PR to reflect above points.